tools/prerequisite-check.sh

```diff
@@ -0,0 +1,104 @@
+#!/bin/bash
+
+REQUIRED_GO_VERSION=1.11
+REQUIRED_SGXSDK_VERSION=2.3.101
+
+show_os_info() {
+	echo "OS: $(. /etc/os-release ; echo $PRETTY_NAME)"
+}
+
+compare_version() {
+	local ver1=$(echo "$1" | awk -F. '{ printf("%d%03d%03d%05d\n", $1,$2,$3,$4); }';)
+	local ver2=$(echo "$2" | awk -F. '{ printf("%d%03d%03d%05d\n", $1,$2,$3,$4); }';)
+
+	if [ "$ver1" -lt "$ver2" ] ; then
+		return 1
+	else
+		return 0
+	fi
+}
+
+check_go_version() {
+	if [[ "$(go version)" =~ go([0-9]+\.[0-9]+\.?[0-9]*) ]] ; then
+		local goversion=${BASH_REMATCH[1]}
+	else
+		echo "Failed to parse go version. Please install it."
+		return 1
+	fi
+
+	if ! compare_version "$goversion" "$REQUIRED_GO_VERSION" ; then
+		echo "Your Go (version $goversion) is older than required version ($REQUIRED_GO_VERSION), so please update it."
+		return 1
+	fi
+
+	echo "Go version: $goversion"
+}
+
+check_sgxsdk_version() {
+	local sgxsdk_version=$(grep ^Version: $SGX_SDK/pkgconfig/libsgx_urts.pc | awk '{print $2}')
+	if [ ! "$sgxsdk_version" ] ; then
+		echo "Failed to parse SGX SDK version from $SGX_SDK/pkgconfig/libsgx_urts.pc. Please make sure that SGX SDK is properly installed."
+		return 1
+	fi
+
+	if ! compare_version $sgxsdk_version $REQUIRED_SGXSDK_VERSION ; then
+		echo "Your SGX SDK (version $sgxsdk_version) is older than required version ($REQUIRED_SGXSDK_VERSION), so please update it."
+		return 1
+	fi
+
+	echo "Intel SGX SDK version: $sgxsdk_version"
+}
+
+check_sgx_support() {
+	if [ "$SGX_MODE" == SIM ] ; then
+		return 0
+	fi
+
+	if make -s --no-print-directory -C tools/sgx-capability check ; then
+		return 0
+	else
+		echo "Intel SGX is not available. Please set environment variable SGX_MODE=SIM."
+		return 1
+	fi
+}
+
+check_sgx_psw_installed() {
+	if [ "$SGX_MODE" == SIM ] ; then
+		return 0
+	fi
+
+	dpkg -l libsgx-enclave-common > /dev/null
```

Comment:
I think it is too strict about the assumption of Ubuntu, because Intel SGX [officially supports](https://01.org/intel-softwareguard-extensions/downloads/intel-sgx-linux-2.3.1-release) not only Ubuntu but also Red Hat Enterprise Linux, CentOS, Fedora, and SUSE Linux Enterprise Server, and our code don't depends on specific operating system.

How about checking existence of `/opt/intel/sgxpsw/uninstall.sh` here not to depend on OS-specific package managers? According to the document ([v2.3.1](https://download.01.org/intel-sgx/linux-2.3.1/docs/Intel_SGX_Installation_Guide_Linux_2.3.1_Open_Source.pdf) and [v2.8](https://download.01.org/intel-sgx/sgx-linux/2.8/docs/Intel_SGX_Installation_Guide_Linux_2.8_Open_Source.pdf) (the latest release), p. 12), Intel SGX PSW installs the script in the path.

Reply:
> How about checking existence of `/opt/intel/sgxpsw/uninstall.sh` here not to depend on OS-specific package managers? According to the document ([v2.3.1](https://download.01.org/intel-sgx/linux-2.3.1/docs/Intel_SGX_Installation_Guide_Linux_2.3.1_Open_Source.pdf) and [v2.8](https://download.01.org/intel-sgx/sgx-linux/2.8/docs/Intel_SGX_Installation_Guide_Linux_2.8_Open_Source.pdf) (the latest release), p. 12), Intel SGX PSW installs the script in the path.

It seems that `/opt/intel/sgxpsw/uninstall.sh` is not installed when installing SGX PSW from Debian repo, so we can't rely on this file as generic solution. Actually, the installation guide mentions the installer `sgx_linux_<os>_x64_psw_<version>.bin` in "Uninstall the Intel®SGX PSW package" section, but that file does not exist, so this paragraph might be outdated.

I thought that we could use `/usr/lib/x86_64-linux-gnu/libsgx_enclave_common.so.1` as a indicator, but that's located differently depending on Linux distributions (for example it's `/usr/lib64/libsgx_enclave_common.so.1` for RHEL). Another thought is to match "libsgx_enclave_common.so" over `find /usr/` or `ldconfig -p` output, although it might look like a hack.

Reply:
Thanks for the info, I have no good idea and agree with this should not be complicated process; it is no problem with proceeding with this implementation.

Reply:
What configure scripts usually do to detect missing dependencies is try building a tiny program (usually inlined in the script) and check if the dependency is resolved. But this may be too complicated for us at this stage.